File: services/nuldoc/nuldoc-src/generators/slide.ts
import { join } from "@std/path";
import SlidePage from "../pages/SlidePage.ts";
import { Config } from "../config.ts";
import { Page } from "../page.ts";
import { Date, Revision } from "../revision.ts";
import { Slide } from "../slide/slide.ts";
import { getPostPublishedDate, getPostUpdatedDate } from "./post.ts";

export interface SlidePage extends Page {
  title: string;
  description: string;
  event: string;
  talkType: string;
  slideLink: string;
  tags: string[];
  revisions: Revision[];
  published: Date;
  updated: Date;
  uuid: string;
}

export async function generateSlidePage(
  slide: Slide,
  config: Config,
): Promise<SlidePage> {
  const html = await SlidePage(slide, config);

  const cwd = Deno.cwd();
  const contentDir = join(cwd, config.locations.contentDir);
  const destFilePath = join(
    slide.sourceFilePath.replace(contentDir, "").replace(".toml", ""),
    "index.html",
  );
  return {
    root: html,
    renderer: "html",
    site: "slides",
    destFilePath: destFilePath,
    href: destFilePath.replace("index.html", ""),
    title: slide.title,
    description: `登壇: ${slide.event} (${slide.talkType})`,
    event: slide.event,
    talkType: slide.talkType,
    slideLink: slide.slideLink,
    tags: slide.tags,
    revisions: slide.revisions,
    published: getPostPublishedDate(slide),
    updated: getPostUpdatedDate(slide),
    uuid: slide.uuid,
  };
}

File: services/nuldoc/lib/nuldoc/generators/slide.rb
module Nuldoc
  SlidePageData = Data.define(:root, :renderer, :site, :dest_file_path, :href,
                              :title, :description, :event, :talk_type, :slide_link,
                              :tags, :revisions, :published, :updated, :uuid)

  module Generators
    class Slide
      def initialize(slide, config)
        @slide = slide
        @config = config
      end

      def generate
        html = Pages::SlidePage.render(slide: @slide, config: @config)

        content_dir = File.join(Dir.pwd, @config.locations.content_dir)
        dest_file_path = File.join(
          @slide.source_file_path.sub(content_dir, '').sub('.toml', ''),
          'index.html'
        )

        SlidePageData.new(
          root: html,
          renderer: :html,
          site: 'slides',
          dest_file_path: dest_file_path,
          href: dest_file_path.sub('index.html', ''),
          title: @slide.title,
          description: "#{@slide.event} (#{@slide.talk_type})",
          event: @slide.event,
          talk_type: @slide.talk_type,
          slide_link: @slide.slide_link,
          tags: @slide.tags,
          revisions: @slide.revisions,
          published: GeneratorUtils.published_date(@slide),
          updated: GeneratorUtils.updated_date(@slide),
          uuid: @slide.uuid
        )
      end
    end
  end
end
